crypto/http/http_client.c

```diff
@@ -733,10 +770,16 @@ static ASN1_VALUE *BIO_mem_d2i(BIO *mem, const ASN1_ITEM *it)
 
     if ((resp = ASN1_item_d2i(NULL, &p, BIO_get_mem_data(mem, &p), it)) == NULL)
         ERR_raise(ERR_LIB_HTTP, HTTP_R_RESPONSE_PARSE_ERROR);
+    BIO_free(mem);
```

Comment:
Please no. This makes the API contract of the function awkward. And I understand it complicates OCSP_REQ_CTX_nbio_d2i() implementation in the sense it will require additional function, but I still think it is worth it rather than creating functions that "use a bio and close it".

Reply:
Hmm, I see your point that freeing the BIO may be unexpected, but as you noticed this very hand because otherwise the caller must use a local variable to store the intermediate result, convert it, and then free the BIO himself.
I think I've found a better solution: as written above, I've renamed the function to `OSSL_HTTP_d2i_free_bio`, which makes crystal clear that the BIO is consumed (as also will be written in the doc, which is not part of this PR).
This way we have bot a telling name and very convenient use - okay?

Reply:
The nice thing about the design of `OSSL_HTTP_i2d_new_bio` and `OSSL_HTTP_d2i_free_bio` is that they can be used, where needed, like pipe transformers, converting between DER-encoded BIOs and the internal C structure representation.

Reply:
The nice "piping" is lost if you try to add proper error checking in between. I suppose you can handle NULL bio on input of these functions as well, but it requires extra care and thought about the error reporting.

Reply:
Also even if we stay with the concept the name `OSSL_HTTP_d2i_free_bio` is less than ideal because it can be interpreted also as "Free a bio that was created by some OSSL_HTTP_d2i_new_bio.".  So maybe better name would be `OSSL_HTTP_d2i_consume_bio` :grin: which is unfortunately awkward as well.

Reply:
To accept this API concept I'd like to see an opinion of others as well.

Reply:
> The nice "piping" is lost if you try to add proper error checking in between. I suppose you can handle NULL bio on input of these functions as well, but it requires extra care and thought about the error reporting.

The function is defined like this:
```
ASN1_VALUE *OSSL_HTTP_d2i_free_bio(BIO *mem, const ASN1_ITEM *it)
{
    const unsigned char *p;
    ASN1_VALUE *resp;

    if (mem == NULL)
        return NULL;

    if ((resp = ASN1_item_d2i(NULL, &p, BIO_get_mem_data(mem, &p), it)) == NULL)
        ERR_raise(ERR_LIB_HTTP, HTTP_R_RESPONSE_PARSE_ERROR);
    BIO_free(mem);
    return resp;
}
```
so this does include NULL handling (which is, as before, simply propagation of any error(s) that occurred before)
and also adds its own error handling and reporting in case the parsing fails.
I do not see anything missing here.
Which further "proper error checking" do you mean?

Reply:
> So maybe better name would be `OSSL_HTTP_d2i_consume_bio` which is unfortunately awkward as well.

I like the new name. Though it is slightly longer, it is even more clear. Changed this way.

Reply:
> Which further "proper error checking" do you mean?

I've meant if the caller wants to discern between errors in the function that creates the bio and errors in the OSSL_HTTP_d2i_consume_bio() he would still have to avoid the nice piping of calls.

Reply:
Also I did not mean you have it implemented wrong, it's just that it's easy to make mistakes in such kind of functions - such as - you always have to free the passed-in bio regardless of errors and you need to do nothing in case the bio is NULL. Which you've correctly done, of course.

Reply:
I see - but I do not see much point in giving the chance for discerning those.
Recall that before, e.g., in `apps.c` the use was like this
```
    resp = OSSL_HTTP_get_asn1(url, proxy, no_proxy,
                              NULL, NULL, app_http_tls_cb, &info,
                              headers, 0 /* maxline */, 0 /* max_resp_len */,
                              timeout, expected_content_type, it);
```
and now it's like this:
```
    resp = OSSL_HTTP_d2i_free_bio(OSSL_HTTP_get(url, proxy, no_proxy,
                                                NULL, NULL,
                                                app_http_tls_cb, &info,
                                                0 /* buf_size */, headers,
                                                expected_content_type,
                                                1 /* expect_asn1 */,
                                                HTTP_DEFAULT_MAX_RESP_LEN,
                                                timeout), it);
```
So formerly the functionality of `OSSL_HTTP_d2i_free_bio` and `OSSL_HTTP_get` was combined in `OSSL_HTTP_get_asn1`, and then caller was not able to differentiate the exact source of any errors that occurred in `OSSL_HTTP_get_asn1`, neither.
And this was not considered a problem, neither.

Reply:
This function shouldn't exist at all, as it duplicates already existing functionality:

https://github.com/openssl/openssl/blob/592ea4ba94b790a9c366fd12792d88fb9c28ef88/include/openssl/asn1.h.in#L787

And while I understand the piping idea, this form makes for patterns that aren't the best.  I would much prefer something like this:

``` C
    b = OSSL_HTTP_get(url, proxy, no_proxy, NULL, NULL, app_http_tls_cb, &info,
                      0 /* buf_size */, headers, expected_content_type, 1 /* expect_asn1 */,
                      HTTP_DEFAULT_MAX_RESP_LEN, timeout);
    resp = ASN1_item_d2i_bio(it, b, NULL);
    BIO_free(b);
```

Reply:
If we take out the `BIO_free(mem);` from `OSSL_HTTP_d2i_consume_bio` and call the remaining function `OSSL_HTTP_d2i_bio`, _each_ use would become just (IMHO needlessly) cumbersome like this:
```
    BIO mem;

    resp = NULL;
    mem = OSSL_HTTP_get(url, proxy, no_proxy,
                        NULL, NULL,
                        app_http_tls_cb, &info,
                        0 /* buf_size */, headers,
                        expected_content_type,
                        1 /* expect_asn1 */,
                        HTTP_DEFAULT_MAX_RESP_LEN,
                        timeout);
    if (mem != NULL) {
        resp = OSSL_HTTP_d2i_bio(mem, it);
        BIO_free(mem);
    }

```

Reply:
1.  Why do you insist on making a duplicate of `ASN1_item_d2i_bio()`?
2.  Both `ASN1_item_d2i_bio()` and `BIO_free()` check that the BIO is non-NULL, or report an error if it is, so all your added checks aren't really necessary.

Point is that your construction as it's currently made up is trading cumbersome for confusing.

Reply:
>     1. Why do you insist on making a duplicate of `ASN1_item_d2i_bio()`?

When I sent my last response I had not yet seen yours
(my browser often does not update GitHub discussion threads).

>     2. Both `ASN1_item_d2i_bio()` and `BIO_free()` check that the BIO is non-NULL, or report an error if it is, so all your added checks aren't really necessary.

Would be nice, but AFAICS `ASN1_item_d2i_bio()` can crash on NULL input.
So I suggest that I make sure that it simply return NULL in this case,
and then indeed `OSSL_HTTP_d2i_bio` is not needed and we can use, as you wrote:
```
    resp = ASN1_item_d2i_bio(it, b, NULL);
    BIO_free(b);
```

Reply:
> Would be nice, but AFAICS `ASN1_item_d2i_bio()` can crash on NULL BiO input.

I suggest you test that.  What I can see is that it calls `BIO_read()`, and in the actual reading function (internal), there's this:

https://github.com/openssl/openssl/blob/f71a7453589b29819f2e35b8cf08c8423b0d27a3/crypto/bio/bio_lib.c#L261-L264

Reply:
Except the added ERR_R_PASSED_NULL_PARAMETER would then just hide the real error. (I know it will be deeper in the err stack but...)

Reply:
Hmm, you are right that `ASN1_item_d2i_bio(it, NULL, ...)` does not crash.
But it produces even two error queue entries which are misleading in this case:
```
ERR_raise(ERR_LIB_BIO, ERR_R_PASSED_NULL_PARAMETER);
ERR_raise(ERR_LIB_ASN1, ASN1_R_NOT_ENOUGH_DATA);
```
so applications still need to use, e.g.,
```
    resp = NULL;
    if (b != NULL)
         resp = ASN1_item_d2i_bio(it, b, NULL);
    BIO_free(b);
```
to prevent those. Pretty ugly.

Ok if I add to 
`void *ASN1_item_d2i_bio(const ASN1_ITEM *it, BIO *in, void *x)`
the following two lines?
```
    if (in == NULL)
        return NULL;
```
This way callers can use as you suggested:
```
    resp = ASN1_item_d2i_bio(it, b, NULL);
    BIO_free(b);
```

Reply:
> Except the added ERR_R_PASSED_NULL_PARAMETER would then just hide the real error. (I know it will be deeper in the err stack but...)

Just saw your comment.
Indeed the error is needless and misleading.
So I'll add the NULL check to `ASN1_item_d2i_bio()`, 
and on this occasion also replace in `asn1_d2i_read_bio()`
the needlessly hard `ERR_clear_error();` by `ERR_set_mark()` and friends.

Reply:
Done.

I also noticed that the other proposed function, `OSSL_HTTP_i2d_new_bio()`,
actually has broader applicability. I renamed it to `ASN1_item_i2d_mem_bio()`, moved it to crypto/asn1/, and documented it, and on this occasion documented also `ASN1_item_d2i_bio().`

Reply:
Doing so I noticed that there was a bug in `OCSP_sendreq_bio()` and that the legacy function `OCSP_sendreq_nbio()` is better expressed by `OSSL_HTTP_REQ_CTX_nbio_d2i()`, so changed these as well.